File: backend/src/models/ServicePlan.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IServicePlan extends Document {
  name: string;
  amountKes: number;
  frequency: 'daily' | 'weekly' | 'monthly' | 'annually';
  billingDay: number; // Day of the month (1-31) or day of the week (1-7) depending on frequency
  ownerId: mongoose.Types.ObjectId;
}

const ServicePlanSchema: Schema = new Schema(
  {
    name: { type: String, required: true, trim: true },
    amountKes: { type: Number, required: true, min: 0 },
    frequency: {
      type: String,
      enum: ['daily', 'weekly', 'monthly', 'annually'],
      required: true,
    },
    billingDay: { type: Number, required: true, min: 1, max: 31 }, // Further validation based on frequency can be added in pre-save hooks or controllers
    ownerId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  },
  { timestamps: true }
);

export default mongoose.model<IServicePlan>('ServicePlan', ServicePlanSchema);

File: backend/app/models/service_plan.py
from datetime import datetime, timezone
from typing import Literal
from beanie import Document, PydanticObjectId
from pydantic import Field


class ServicePlan(Document):
    name: str
    amount_kes: float = Field(gt=0, alias="amountKes")
    frequency: Literal["daily", "weekly", "monthly", "quarterly", "annually"]
    billing_day: int = Field(ge=1, le=31, alias="billingDay")
    owner_id: PydanticObjectId = Field(alias="ownerId")
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "serviceplans"
        populate_by_name = True
        indexes = [
            "owner_id",
            "name",
        ]
